presto-spi/src/main/java/com/facebook/presto/spi/block/VariableWidthBlockBuilder.java

```diff
@@ -300,8 +300,18 @@ public Block copyRegion(int positionOffset, int length)
         }
 
         int[] newOffsets = Arrays.copyOfRange(offsets, positionOffset, positionOffset + length + 1);
+        // set new offsets to start from beginning of slice (since we are copying)
+        for (int i = 0; i < newOffsets.length; i++) {
+            newOffsets[i] -= offsets[positionOffset];
+        }
         boolean[] newValueIsNull = Arrays.copyOfRange(valueIsNull, positionOffset, positionOffset + length);
-        return new VariableWidthBlock(length, sliceOutput.slice(), newOffsets, newValueIsNull);
+
+        Slice slice = sliceOutput.slice().slice(offsets[positionOffset], newOffsets[length]);
```

Comment:
We can also always create a new Slice, like what we did in `FixedWidthBlockBuilder`: 

```
        Slice slilce = Slices.copyOf(sliceOutput.getUnderlyingSlice(), offsets[positionOffset], newOffsets[length]);
```

Reply:
Now that we have the option to avoid unnecessarily creating slices, we should take advantage of it.